### ISSUE TITLE
Load Tests Against Azure Environment

### ISSUE THREAD

**Issue by JohnNKing** [CDCgov/trusted-intermediary#1122]
# DevEx/OpEx

Currently, load tests are only run locally. Obtain more realistic data on TI service performance by running the tests within Azure.

## Tasks

- [ ] Something.

## Additional Context

Add any other context or screenshots about the work here.

**Comment by jcrichlake** [comment]
In order to utilize our existing load tests we need to sign up for the azure locust preview test program.  This will take a few days to be completed. 

Adding the blocked tag to this until we hear back

**Comment by brick-green** [comment]
@jcrichlake Who are we waiting to hear back from?

**Comment by halprin** [comment]
we got past the current issue of the load tests.  The problem was we were missing `https://` (i.e. the protocol) for `LOCUST_HOST`.  The load test successfully hit Internal.

We can probably start adding back the other parts of the load test.


We also need to figure out how to authenticate agains the `token` endpoint.


There seems to be a section for secrets in the load test config, so perhaps we could use it to fill in a JWT or perhaps the private key that is used to sign JWTs.

Updated this ticket with additional tasks and context.

**Comment by halprin** [comment]
I got the original load test file working in Azure (with some modifications) which includes authenticating against the `token` endpoint and hitting the other endpoints.  I got the secrets working with the JWT.

We've only ever done short load tests, so I will experiment with longer load tests and make sure the Azure load test framework continues to work.

**Comment by halprin** [comment]
@somesylvie, is the task `Look into why failing to link metadata results in 5XXs.` complete since you fixed our database in Internal?  Your load test and my subsequent larger load test seems to have 0 errors.

**Comment by halprin** [comment]
@somesylvie, I ran a larger test after your latest one (description of `Bigger test`).  I ran 2 engines, each with 50 users, for 10 minutes.  It passed with flying colors.

What are your thoughts on marking the task `Optional: look into why our deployed environments are slow for the orders and results endpoint (is it really due to RS staging being slow?).` complete?  Based on your last test and my larger test, the slowness _seems_ to be due to RS.  Thoughts?

**Comment by somesylvie** [comment]
> @somesylvie, is the task `Look into why failing to link metadata results in 5XXs.` complete since you fixed our database in Internal? Your load test and my subsequent larger load test seems to have 0 errors.

Yep, I think this task is completed/resolved

**Comment by somesylvie** [comment]
> @somesylvie, I ran a larger test after your latest one (description of `Bigger test`). I ran 2 engines, each with 50 users, for 10 minutes. It passed with flying colors.
> 
> What are your thoughts on marking the task `Optional: look into why our deployed environments are slow for the orders and results endpoint (is it really due to RS staging being slow?).` complete? Based on your last test and my larger test, the slowness _seems_ to be due to RS. Thoughts?

Yeah, this seems plausible. If we start running into timing issues in future, we can assess them then

**Comment by pluckyswan** [comment]
We are closing out [an existing card regarding load tests against staging](https://github.com/CDCgov/trusted-intermediary/issues/628). Can this be included in this card's scope: `We may need to re-examine our load test assumptions, we are currently ramping up to 1k users and then hammering the various endpoints, this may be an amount of load that is either far off in the future or will require multiple instances of the intermediary running to handle it.`

**Comment by halprin** [comment]
> We are closing out [an existing card regarding load tests against staging](https://github.com/CDCgov/trusted-intermediary/issues/628). Can this be included in this card's scope: `We may need to re-examine our load test assumptions, we are currently ramping up to 1k users and then hammering the various endpoints, this may be an amount of load that is either far off in the future or will require multiple instances of the intermediary running to handle it.`

Definitely, it will be included.  We'll be figuring out what is a good level of load now that we are hitting real environments.  Whenever your changing the load test to hit a different environment (in this case, staging versus local), one should always re-evaluate what is a good, realistic traffic load.

**Comment by jcrichlake** [comment]
Two questions I have on this card.  

Does Flexion need to research a way to charge the CDC for the cost associated with the load testing? Pricing can be found at the below link:

https://azure.microsoft.com/en-us/pricing/details/load-testing/

**Comment by brick-green** [comment]
Cannot charge to CDC as this is in Internal flexion infrastructure

**Comment by jcrichlake** [comment]
@brick-green It is Flexion infrastructure, but the Flexion infrastructure is solely being used to load test containers in the CDC Azure environment. The reason we can't do this in the CDC Azure is because we are using a subscription wide opt in to enable Locust load testing.

**Comment by halprin** [comment]
[PR](https://github.com/CDCgov/trusted-intermediary/pull/1665) was merged.  I will move the work to get this set-up in Staging into another DevEx/OpEx ticket.
